state/etcd/etcd.go

```diff
@@ -0,0 +1,499 @@
+		return nil, fmt.Errorf("couldn't get key %s: %w", keyWithPath, err)
+	}
+
+	if resp == nil || len(resp.Kvs) == 0 {
+		return &state.GetResponse{}, nil
+	}
+
+	return &state.GetResponse{
+		Data: resp.Kvs[0].Value,
+		ETag: ptr.Of(strconv.Itoa(int(resp.Kvs[0].ModRevision))),
+	}, nil
+}
+
+// Set saves a Etcd KV item.
+func (e *Etcd) Set(ctx context.Context, req *state.SetRequest) error {
+	ttlInSeconds, err := e.doSetValidateParameters(req)
+	if err != nil {
+		return err
+	}
+
+	keyWithPath := e.keyPrefixPath + "/" + req.Key
+	err = e.doValidateEtag(keyWithPath, req.ETag, req.Options.Concurrency)
+	if err != nil {
+		return err
+	}
+
+	reqVal, err := utils.Marshal(req.Value, json.Marshal)
+	if err != nil {
+		return err
+	}
+
+	return e.doSet(ctx, keyWithPath, string(reqVal), req.ETag, ttlInSeconds)
+}
+
+func (e *Etcd) BulkSet(ctx context.Context, req []state.SetRequest) error {
+	if len(req) == 0 {
+		return nil
+	}
+
+	for _, dr := range req {
+		ttlInSeconds, err := e.doSetValidateParameters(&dr)
+		if err != nil {
+			return err
+		}
+
+		keyWithPath := e.keyPrefixPath + "/" + dr.Key
+		err = e.doValidateEtag(keyWithPath, dr.ETag, dr.Options.Concurrency)
+		if err != nil {
+			return err
+		}
+
+		reqVal, err := utils.Marshal(dr.Value, json.Marshal)
+		if err != nil {
+			return err
+		}
+
+		err = e.doSet(ctx, keyWithPath, string(reqVal), dr.ETag, ttlInSeconds)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (e *Etcd) doSet(ctx context.Context, key, reqVal string, etag *string, ttlInSeconds int64) error {
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	if ttlInSeconds > 0 {
+		resp, err := e.client.Grant(ctx, ttlInSeconds)
+		if err != nil {
+			return fmt.Errorf("couldn't grant lease %s: %w", key, err)
+		}
+		if etag != nil {
+			etag, _ := strconv.ParseInt(*etag, 10, 64)
+			_, err = e.client.Txn(ctx).
+				If(clientv3.Compare(clientv3.ModRevision(key), "=", etag)).
+				Then(clientv3.OpPut(key, reqVal, clientv3.WithLease(resp.ID))).
+				Commit()
+		} else {
+			_, err = e.client.Put(ctx, key, reqVal, clientv3.WithLease(resp.ID))
+		}
+		if err != nil {
+			return fmt.Errorf("couldn't set key %s: %w", key, err)
+		}
+	} else {
+		var err error
+		if etag != nil {
+			etag, _ := strconv.ParseInt(*etag, 10, 64)
+			_, err = e.client.Txn(ctx).
+				If(clientv3.Compare(clientv3.ModRevision(key), "=", etag)).
+				Then(clientv3.OpPut(key, reqVal)).
+				Commit()
+		} else {
+			_, err = e.client.Put(ctx, key, reqVal)
+		}
+		if err != nil {
+			return fmt.Errorf("couldn't set key %s: %w", key, err)
+		}
+	}
+	return nil
+}
+
+func (e *Etcd) doSetValidateParameters(req *state.SetRequest) (int64, error) {
+	err := state.CheckRequestOptions(req.Options)
+	if err != nil {
+		return 0, err
+	}
+
+	ttlInSeconds, err := doParseTTLInSeconds(req.Metadata)
+	if err != nil {
+		return 0, err
+	}
+
+	return ttlInSeconds, nil
+}
+
+func doParseTTLInSeconds(metadata map[string]string) (int64, error) {
```

Comment:
Please use the method from the stateutils package: https://github.com/windnotstop/components-contrib/blob/fad106b3fb89cc65ee2d8d5592613bc62b80863d/state/utils/ttl.go